src/include/network/connection_handle.h

```diff
@@ -0,0 +1,227 @@
+  }
+
+  // Exposed for testing
+  const std::unique_ptr<ProtocolHandler> &GetProtocolHandler() const {
+    return protocol_handler_;
+  }
+
+  // State Machine actions
+  /**
+   * refill_read_buffer - Used to repopulate read buffer with a fresh
+   * batch of data from the socket
+   */
+  Transition FillReadBuffer();
+  Transition Wait();
+  Transition Process();
+  Transition ProcessWrite();
+  Transition GetResult();
+  Transition CloseSocket();
+
+ private:
+  /**
+   * A state machine is defined to be a set of states, a set of symbols it
+   * supports, and a function mapping each
+   * state and symbol pair to the state it should transition to. i.e.
+   * transition_graph = state * symbol -> state
+   *
+   * In addition to the transition system, our network state machine also needs
+   * to perform actions. Actions are
+   * defined as functions (lambdas, or closures, in various other languages) and
+   * is promised to be invoked by the
+   * state machine after each transition if registered in the transition graph.
+   *
+   * So the transition graph overall has type transition_graph = state * symbol
+   * -> state * action
+   */
+  class StateMachine {
+   public:
+    using action = Transition (*)(ConnectionHandle &);
+    using transition_result = std::pair<ConnState, action>;
+    /**
+     * Runs the internal state machine, starting from the symbol given, until no
+     * more
+     * symbols are available.
+     *
+     * Each state of the state machine defines a map from a transition symbol to
+     * an action
+     * and the next state it should go to. The actions can either generate the
+     * next symbol,
+     * which means the state machine will continue to run on the generated
+     * symbol, or signal
+     * that there is no more symbols that can be generated, at which point the
+     * state machine
+     * will stop running and return, waiting for an external event (user
+     * interaction, or system event)
+     * to generate the next symbol.
+     *
+     * @param action starting symbol
+     * @param connection the network connection object to apply actions to
+     */
+    void Accept(Transition action, ConnectionHandle &connection);
+
+   private:
+    /**
+     * delta is the transition function that defines, for each state, its
+     * behavior and the
+     * next state it should go to.
+     */
+    static transition_result Delta_(ConnState state, Transition transition);
+    ConnState current_state_ = ConnState::READ;
+  };
+
+  friend class StateMachine;
+  friend class ConnectionHandleFactory;
+
+  ConnectionHandle(int sock_fd, ConnectionHandlerTask *handler,
+                   std::shared_ptr<Buffer> rbuf, std::shared_ptr<Buffer> wbuf);
+
+  ProcessResult ProcessInitial();
+
+  /**
+   * Extracts the header of a Postgres start up packet from the read socket
+   * buffer
+   */
+  static bool ReadStartupPacketHeader(Buffer &rbuf, InputPacket &rpkt);
```

Comment:
This function would need to be put back to protocol handler. I think the SSL pr has done some of this.

Reply:
We should probably resolve these when we merge that PR?